### PR TITLE
Add interceptor method to fail hijacked send messages and get status on POST_SEND_MESSAGE

### PR DESCRIPTION
**Pull request by yashykt**
1) Add interceptor method to fail hijacked send messages
2) Add interception point POST_SEND_MESSAGE
3) Add interceptor method to check for success of send message op

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.2%    +733 [None]                                                                               +2.10Ki  +0.0%
      +0.3%    +717 [Unmapped]                                                                           +2.08Ki  +0.0%
      +5.9%      +8 vtable for grpc::internal::InterceptorBatchMethodsImpl                                    +8  +5.9%
      +5.9%      +8 vtable for grpc::internal::CancelInterceptorBatchMethods                                  +8  +5.9%
  +0.8%    +398 src/cpp/server/server_cc.cc                                                             +398  +0.8%
      [NEW]    +200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +200  [NEW]
      +7.7%    +124 grpc::Server::Server                                                                    +124  +7.7%
      [NEW]     +92 grpc::experimental::ClientRpcInfo::RunInterceptor                                        +92  [NEW]
      +5.1%     +64 grpc::Server::SyncRequest::CallData::ContinueRunAfterInterception                        +64  +5.1%
      +2.0%     +48 grpc::Server::UnimplementedAsyncRequest::FinalizeResult                                  +48  +2.0%
      +2.2%     +48 grpc::Server::UnimplementedAsyncRequest::UnimplementedAsyncRequest                       +48  +2.2%
      +1.2%     +48 grpc::Server::Start                                                                      +48  +1.2%
      +3.8%     +16 grpc::ServerInterface::BaseAsyncRequest::BaseAsyncRequest                                +16  +3.8%
      +0.5%     +16 grpc::Server::RegisterService                                                            +16  +0.5%
      +0.8%     +16 grpc::Server::SyncRequestThreadManager::DoWork                                           +16  +0.8%
      +0.8%     +16 grpc::internal::ErrorMethodHandler<(grpc::StatusCode)12>::RunHandler                     +16  +0.8%
      +0.8%     +16 grpc::internal::ErrorMethodHandler<(grpc::StatusCode)8>::RunHandler                      +16  +0.8%
      +1.5%     +10 [Unmapped]                                                                               +10  +1.5%
  +0.7%    +313 src/cpp/client/generic_stub.cc                                                          +313  +0.7%
      [NEW]    +200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +200  [NEW]
      [NEW]     +54 grpc::internal::InterceptorBatchMethodsImpl::FailHijackedSendMessage                     +54  [NEW]
      +2.3%     +48 grpc::GenericStub::PrepareCall                                                           +48  +2.3%
      +2.0%     +48 grpc::GenericStub::Call                                                                  +48  +2.0%
      +4.6%     +32 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +32  +4.6%
      +4.3%     +24 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +24  +4.3%
      +4.3%     +24 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +24  +4.3%
      +1.0%     +16 grpc::GenericStub::PrepareUnaryCall                                                      +16  +1.0%
      +0.5%     +16 grpc::GenericStub::experimental_type::UnaryCall                                          +16  +0.5%
      +0.6%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +0.6%
      +3.2%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +3.2%
      +0.6%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +0.6%
      +2.7%      +3 grpc::internal::InterceptorBatchMethodsImpl::GetSendStatus                                +3  +2.7%
  +0.8%     +94 src/cpp/server/server_context.cc                                                         +94  +0.8%
      [NEW]    +200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +200  [NEW]
      +4.4%     +30 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +30  +4.4%
      +3.0%     +24 grpc::ServerContext::ServerContext                                                       +24  +3.0%
      +1.9%     +16 grpc::ServerContext::ServerContext                                                       +16  +1.9%
      +3.3%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +3.3%
      +0.9%      +8 grpc::ServerContext::BeginCompletionOp                                                    +8  +0.9%
  +0.7%     +38 src/cpp/client/client_context.cc                                                         +38  +0.7%
      [NEW]     +38 grpc::internal::CancelInterceptorBatchMethods::FailHijackedSendMessage                   +38  [NEW]

 -------------- SHRINKING                                                                            --------------
  -0.0%     -16 src/cpp/server/health/default_health_check_service.cc                                    -16  -0.0%
      [DEL]    -244 std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<c    -244  [DEL]
      [DEL]    -200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    -200  [DEL]
     -21.2%     -89 grpc::DefaultHealthCheckService::HealthCheckServiceImpl::WatchCallHandler::SendFinis     -89 -21.2%
      -6.2%     -15 [Unmapped]                                                                               -15  -6.2%

  +0.3% +1.52Ki TOTAL                                                                                +2.91Ki  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,964      Total (=)      2,019,964

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,183,084      Total (<)     11,183,091

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,020,494      Total (=)      2,020,494

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,175,631      Total (>)     11,175,628

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,020,494      Total (=)      2,020,494

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,175,632      Total (>)     11,175,631

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,020,494      Total (=)      2,020,494

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,175,631      Total (>)     11,175,625

 No significant differences in binary sizes


```

**Comment by yashykt** [comment]
Reviving this PR

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  +0.3%    +170 src/cpp/client/generic_stub.cc                                                          +170  +0.3%
      +152%    +476 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:    +476  +152%
      [NEW]    +200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +200  [NEW]
      [NEW]     +54 grpc::internal::InterceptorBatchMethodsImpl::FailHijackedSendMessage                     +54  [NEW]
      +2.3%     +48 grpc::GenericStub::PrepareCall                                                           +48  +2.3%
      +1.7%     +48 grpc::GenericStub::experimental_type::PrepareBidiStreamingCall                           +48  +1.7%
      +2.0%     +48 grpc::GenericStub::Call                                                                  +48  +2.0%
      +2.0%     +32 grpc::GenericStub::PrepareUnaryCall                                                      +32  +2.0%
      +4.3%     +24 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +24  +4.3%
      +4.3%     +24 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +24  +4.3%
      +0.5%     +16 grpc::GenericStub::experimental_type::UnaryCall                                          +16  +0.5%
      +3.2%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +3.2%
      [NEW]     +11 grpc::internal::InterceptorBatchMethodsImpl::GetSendMessageStatus                        +11  [NEW]
      +2.7%      +3 grpc::internal::InterceptorBatchMethodsImpl::GetSendStatus                                +3  +2.7%
  +1.1%    +134 src/cpp/server/server_context.cc                                                        +134  +1.1%
      [NEW]    +200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    +200  [NEW]
      +7.7%     +54 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +54  +7.7%
      +3.0%     +24 grpc::ServerContext::ServerContext                                                       +24  +3.0%
      +1.6%     +16 grpc::ServerContext::BeginCompletionOp                                                   +16  +1.6%
      +1.9%     +16 grpc::ServerContext::ServerContext                                                       +16  +1.9%
      +3.3%     +16 grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal:     +16  +3.3%
      +5.1%      +8 [Unmapped]                                                                                +8  +5.1%
  +1.6%     +84 src/cpp/client/client_context.cc                                                         +84  +1.6%
      [NEW]     +46 grpc::internal::CancelInterceptorBatchMethods::GetSendMessageStatus                      +46  [NEW]
      [NEW]     +38 grpc::internal::CancelInterceptorBatchMethods::FailHijackedSendMessage                   +38  [NEW]

 -------------- SHRINKING                                                                            --------------
  -0.6%    -274 src/cpp/server/server_cc.cc                                                             -274  -0.6%
     -52.1%    -465 grpc::ServerAsyncReaderWriter<grpc::ByteBuffer, grpc::ByteBuffer>::WriteAndFinish       -465 -52.1%
     -11.0%    -215 grpc::Server::SyncRequestThreadManager::DoWork                                          -215 -11.0%
      [DEL]    -200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    -200  [DEL]
      -8.0%     -99 grpc::Server::~Server                                                                    -99  -8.0%
  -0.2%     -90 src/cpp/server/health/default_health_check_service.cc                                    -90  -0.2%
     -11.5%    -260 grpc::DefaultHealthCheckService::HealthCheckServiceImpl::WatchCallHandler::SendHealt    -260 -11.5%
      [DEL]    -244 std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<c    -244  [DEL]
      [DEL]    -200 void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >    -200  [DEL]
      -1.7%      -4 [Unmapped]                                                                                -4  -1.7%

 -+-+-+-+-+-+-+ MIXED                                                                                +-+-+-+-+-+-+-
  +0.4% +1.21Ki [None]                                                                               -3.30Ki  -0.0%
       +12%     +16 vtable for grpc::internal::InterceptorBatchMethodsImpl                                   +16   +12%
       +12%     +16 vtable for grpc::internal::CancelInterceptorBatchMethods                                 +16   +12%

  +0.2% +1.23Ki TOTAL                                                                                -3.28Ki  -0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/cpp/end2end/client_interceptors_end2end_test.cc [hunk 1]
```diff
@@ -270,6 +270,60 @@ class HijackingInterceptorMakesAnotherCallFactory
...
+      methods->Proceed();
+    }
+  }
+
+  static bool GotFailedSend() { return got_failed_send_; }
+
+ private:
+  experimental::ClientRpcInfo* info_;
+  int count_ = 0;
+  static bool got_failed_send_;
```
[the remark this answers is not in the document]
Comment (yashykt) [reply test/cpp/end2end/client_interceptors_end2end_test.cc]: Well, I needed a quick way of checking this for an interceptor. Making it static was the easiest way I could think of.

File: include/grpcpp/impl/codegen/interceptor.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ enum class InterceptionHookPoints {
   /// The first two in this list are for clients and servers
```
Comment (vjpai) [inline include/grpcpp/impl/codegen/interceptor.h]: The first three?

Comment (yashykt) [reply include/grpcpp/impl/codegen/interceptor.h]: oh yeah

**Comment by yashykt** [comment]
Bazel errors look like transient infrastructure issues

**Pull request merged**